### ISSUE TITLE
Feature request: minimal idle footprint — adaptive idle polling and optional lazy load / idle unload

### ISSUE THREAD

**Issue by jroth1111** [raullenchai/vllm-mlx#265]
## Summary

Reduce the idle resource footprint of `rapid-mlx serve` so the OpenAI-compatible
HTTP server can stay running indefinitely on Apple Silicon without consuming
appreciable CPU or memory between requests. Two complementary tiers, either of
which is independently useful:

1. **Light tier — adaptive idle polling.** Make the engine step interval
   adaptive (or event-driven) so the async loop doesn't poll at 1 kHz when
   there is no work. Cheap change, no API surface needed.
2. **Heavy tier — request-lazy load and idle unload.** Add `--lazy-load` and
   `--idle-unload-seconds N` so model weights are loaded only when inference
   is actually requested, and released after configurable idleness.

Goal is *minimal* idle footprint, not literally zero. A small base-server RSS
and well under one full core of CPU when idle is fine.

## Motivation

Local Apple Silicon users often want a stable endpoint always available at
e.g. `http://127.0.0.1:8000/v1` for editors, agents, and OpenAI-compatible
clients. Today, keeping `rapid-mlx serve` running keeps the model resident
and the engine loop polling, even when no requests arrive for hours. On a
36 GB Mac that's roughly ~10 GB Metal active for a Qwen2.5-7B-1M-4bit setup
plus continuous low-rate polling — fine for active sessions, expensive for
intermittent use.

### Idle CPU — engine loop polls at 1 ms

`vllm_mlx/engine_core.py` defines:

```python
@dataclass
class EngineConfig:
    ...
    step_interval: float = 0.001  # 1ms between steps
```

and the run loop is:

```python
while self._running:
    if self.scheduler.has_requests():
        output = await loop.run_in_executor(_executor, self.scheduler.step)
        ...
    else:
        await asyncio.sleep(step_interval)
```

So with no requests in flight, the loop wakes ~1000×/s. `step_interval` is
not exposed as a CLI flag.

### Idle RAM — eager load, no unload path

- `serve_command()` calls `load_model(...)` before `uvicorn.run(...)`.
- `server.load_model(...)` constructs the global `BatchedEngine`.
- FastAPI lifespan startup calls `await _engine.start()`.
- `BatchedEngine.start()` → `_start_llm()` loads weights/tokenizer through
  `load_model_with_fallback(...)` on the dedicated `mlx-step` executor.
- `BatchedEngine.stop()` already tears down the engine, clears
  model/tokenizer/processor, sets `_loaded = False` — but this only runs at
  process shutdown.
- `vllm_mlx/worker.py` ships explicit `sleep` / `wake_up` stubs that no-op
  with the comment "Sleep mode not applicable for MLX (unified memory)".
  That is correct for the vLLM page-out semantics it inherited, but not for
  the user-facing goal of "free the weights from RAM until next request."

Ingredients for unload exist; an idle lifecycle does not.

### Tier 1 — adaptive step interval (small PR)

Replace the fixed 1 ms `asyncio.sleep(step_interval)` with one of:

- An asyncio `Event` set by `add_request(...)` and cleared after the queue
  drains. The idle branch becomes `await event.wait()` with a long timeout
  for periodic housekeeping.
- A two-speed interval: `step_interval` while requests are in flight,
  `idle_step_interval` (e.g. 100 ms or 1 s, configurable) when scheduler
  is empty. Single-line change, preserves current latency for active loads.

Either approach should drop idle CPU by ~50–500× without affecting
first-token latency on a request that arrives mid-idle.

### Tier 2 — lazy load + idle unload

```bash
rapid-mlx serve MODEL \
  --lazy-load \
  --idle-unload-seconds 1800
```

Semantics:

- `--lazy-load`: bind the HTTP server without loading model weights.
- `--idle-unload-seconds N`: after N seconds with no active generations,
  optionally save prefix cache, stop the engine, drop model references,
  and return to standby. `0` means never unload once loaded.
- `/health` stays alive while unloaded; `/health/ready` distinguishes
  `standby`, `loading`, `ready`, `unloading`, `error`.
- `/v1/models` continues returning the configured served model name in
  standby (lifecycle metadata in non-OpenAI health endpoints).

## Implementation constraints

The current code has careful MLX thread-affinity handling that any lazy
implementation must preserve.

Specifically, `_start_llm()` creates a dedicated `ThreadPoolExecutor`
(`thread_name_prefix="mlx-step"`) and loads the model on that worker before
handing the *same* executor to `AsyncEngineCore`. This is important because
`mlx-lm` 0.31+ binds module-level `generation_stream` to the thread that
first triggers it; touching weights on the asyncio loop thread leaks a
stray `Stream(gpu, 1)` that breaks every later eval (#170). A lazy reload
should re-enter the same `BatchedEngine.start()` path behind a lifecycle
lock rather than introducing an alternate load path.

Suggested internal shape:

- A lifecycle manager holding the configured model/scheduler/serve params.
- Inference routes call `await ensure_loaded()` under an async lifecycle
  lock; concurrent first requests await one shared load task.
- Active-request counter so idle unload never runs mid-generation.
- On idle timeout: save prefix cache (if `--enable-prefix-cache`), then
  `await engine.stop()`, then `mx.clear_cache()`.
- Streaming cancellation, client disconnect, and failed loads must always
  decrement the counter and return state to `standby` or `error`, never
  stick in `loading`/`active`.

## Acceptance criteria

**Tier 1:**

- With no in-flight requests, server-process CPU is well under 1% of one
  core (versus the current ~kHz polling cost).
- First-token latency for a request arriving mid-idle is unchanged within
  noise.

**Tier 2:**

- `--lazy-load` binds the configured port without loading model weights;
  RSS is at base-server overhead, not model-resident.
- First inference request loads the model and returns a valid response.
- Concurrent first requests share one load operation.
- After `--idle-unload-seconds` with no active requests, the engine
  unloads and RSS drops back below the resident-model size.
- A later inference request reloads the model and succeeds.
- Streaming cancellation/disconnect properly decrements active tracking.
- Prefix cache save/restore behavior is explicit and matches the
  configured `--enable-prefix-cache` value.
- `/health/ready` clearly distinguishes
  `standby` / `loading` / `ready` / `unloading` / `error`.

## Why this matters

This makes Rapid-MLX comfortable to run as an always-on local inference
service for editors, agents, and OpenAI-compatible clients on
memory-constrained Macs. Tier 1 alone turns the server into something that
can be left running for days at near-zero cost; Tier 2 is what users on
8–16 GB Macs need to keep a stable endpoint without competing with their
foreground work.

**Comment by raullenchai** [comment]
Closing this — **Tier 1 shipped in #280, available since v0.6.23**.

> Idle wake rate: 1000 Hz → 0.2 Hz, first-token latency unchanged.

The engine loop now blocks on an `asyncio.Event` set inside `add_request`, with a 5-second housekeeping timeout for non-`add_request` state changes (prefix cache reload, deep_reset). Verified by 4 new unit tests in `tests/test_idle_event_wakeup.py`.

**Tier 2** (`--lazy-load` and `--idle-unload-seconds N`) was explicitly deferred because it needs careful lifecycle locking around MLX's `generation_stream` thread-affinity — moving it to its own tracker so it doesn't get lost: **#319**.

Thanks for the detailed write-up — the source-level analysis (1 ms `step_interval`, eager `load_model` before lifespan, no unload path) made the Tier 1 fix trivial to design.

The unrelated comment on this thread is being handled separately as part of the May 2026 spam wave; it's not signal on this feature request.

**Comment by raullenchai** [comment]
Thanks @jroth1111 for opening this — feedback from people using and building
with rapid-mlx helps shape what we improve next. 🙏

Beyond this issue, what's the roughest edge you've hit with rapid-mlx?
Reply here, or drop it in our feedback thread →
https://github.com/raullenchai/Rapid-MLX/discussions/1205

Also curious: what are you using rapid-mlx for, and what does your setup
look like? Even a couple of lines would help us prioritize the right things.

If you'd rather chat 1:1, happy to connect by email or whatever channel
is easiest. And if you ever write up your setup (blog/gist/tweet), ping me
and I'll amplify it.
